### DISCUSSION TITLE
Using `ridgeplot` for `sktime` and `skpro` distributional predictions?

### DISCUSSION THREAD

**Discussion by fkiraly** [tpvasconcelos/ridgeplot#173] [General]
For a while I have now been thinking about what a good plotting modality would be for fully distributional predictions, i.e., the output of `predict_proba` in `sktime` or `skpro`.

The challnge is that you have a (marginal) distribution for each entry in a `pandas`-like table, which seems hard to visualize. I've experimented with panels (`matplotlib.subplots`) but I wasn't quit happy with the result.

Now, by accident (just curious clicking), I've discovered `ridgeplot`.

What would you think of using the look & feel of `ridgeplot` as a plotting function in `BaseDistribution`? Where rows are the rows of the data-frame like stucture, and mayb there are also columns (but I am happy with the single-variable case too)

The main difference is that the distribution does not need to be estimated via KDE, you already have it in a form where you can access `pdf`, `cdf`, etc, completely, and you have the quantile function too which helps with selecting x-axis range.

Plotting `cdf` and other distribution defining functions would also be neat, of course `pdf` (if exists), or `cdf` (for survival) are already great.

Imagined usage, sth like

```python
fcst = BuildSth(Complex(params), more_params)
fcst.fit(y, fh=range(1, 10)
y_dist = fcst.predict_proba()

y_dist.plot() # default is pdf for continuous distirbutions
y_dist.plot("cdf")
```

Dependencies-wise, one could imagine `ridgeplot` as a plotting softdep like `matplotlib` or `seaborn`, of `skpro` and therefore indirectly of `sktime`.

What do you think?

**Comment by tpvasconcelos** [comment]
Hey @fkiraly happy to see you here! Give me some time to take a look at this one as I'm not super familiar with that class of forecasters from sktime.

> The main difference is that the distribution does not need to be estimated via KDE, you already have it in a form where you can access pdf, cdf, etc, completely, and you have the quantile function too which helps with selecting x-axis range.

If I understood correctly, this is completely fine when it comes to interacting with ridgeplot, since we also accept x-y traces as input -- bypassing the KDE step altogether

**Comment by fkiraly** [comment]
> as I'm not super familiar with that class of forecasters from sktime.

Most popular forecasters have a probabilistic prediction mode, and they can be filtered by the tag `capability:pred_int`.

Check it out in the forecasting tutorial, the main tutorial of `skpro` also explains the tabular (non-time-series) interfaces arorund it.

> If I understood correctly, this is completely fine when it comes to interacting with ridgeplot, since we also accept x-y traces as input -- bypassing the KDE step altogether

Yes - in the conceptual space of `sktime` / `skpro`, the KDE step is an estimator, of type "distribution from sample". In the desired plot, we would go directly from distribution to plot, without th first step of going to distribution from sample.

**Reply by fkiraly** [reply]
any opinions yet?

**Reply by fkiraly** [reply]
?

**Reply by tpvasconcelos** [reply]
Hey @fkiraly. I've been trying to make some time to look into this issue but unfortunately haven't been able to make the time yet. Is this something that is high on your priority list or something that is currently blocking your work? If so, I can try prioritising it. If not, let me know as well and I will take a look at it with more time later on. Thanks!

**Reply by fkiraly** [reply]
no urgency, just sth that imo would be very nice - whenever you find time.

If in the meanwhile I can help with sth `ridgeplot`, let me know.
How about this? https://github.com/tpvasconcelos/ridgeplot/issues/116

**Comment by tpvasconcelos** [comment]
@fkiraly Just trying to get some early feedback from you. I just tried to map this output from sktime's `01b_forecasting_proba.ipynb` notebook:

<img width="1040" alt="image" src="https://github.com/tpvasconcelos/ridgeplot/assets/17701527/10eff33d-1027-4098-990a-f3075f55979c">

into a ridgeline plot of PDFs (for the same forecast horizon):

<img width="787" alt="image" src="https://github.com/tpvasconcelos/ridgeplot/assets/17701527/0396d46c-1e3e-4d13-b33d-45b53cb50975">

👆 this is what I was able to hack together while trying to figure out how the `BaseDistribution` interface works.

Let me know if this is anything like what you had envisioned, otherwise, could you share a more concrete example I can play around with. A sketch would help.

**Reply by tpvasconcelos** [reply]
First thing I thought looking at this is that it's strange to read a time series vertically. The second question is whether time flows from top to bottom (like in the plot above) or bottom-up...

**Reply by fkiraly** [reply]
hm, very interesting!!

This is exactly what I had in mind, although I wonder whether it can be improved, I'm not 100% sure if this is the best way to plot, as you say. I'm still playing around with the best way to plot a sequence of distributions.

### top-to-bottom vs bottom-up

Yes, I'm wondering that too.

Pro top-to-bottom:

* same order as `DataFrame` samples and index
* same order as reading

Pro bottom-up:

* viewer perspective typically has nearer = closer, that would be bottom

Other thoughts:

* other option: left/right? might also be confusing, but time is typically left/right.
* The orientation could be an option, with top-down default?
* we also should consider the case where the row dimension is *not* time, simply a sample index, like in sktime. E.g., a user might like to compare ridgeplots for group analysis or similar.

### colours

Not sure whether I understand what's going on with the colour coding here.

Using colours might help in general, can you explain?

### other thoughts

* I was wondering whether 5/25/50/75/95 quantiles could be connected by straights, whether that would make things look more "connected", or more confusing.
* a substantially different plot - not ridgeplot - I was wondering about is left/right, plotting multiple quantiles, 5/25/50/75/95. That tends to look overcrowded though, and is not truly a distribution plot.

**Reply by tpvasconcelos** [reply]
## top-to-bottom vs bottom-up

After looking through some real examples online I think that the bottom-up convention feels more natural to the eye when interpreting these are forecast horizons or when following an arrow of time.

See [my comment](https://github.com/tpvasconcelos/ridgeplot/discussions/173#discussioncomment-8566883) where I collected some visual examples:

> Granted that this is not a very significant sample but just for reference: 
> - 2/5 use the bottom-up convention 👆 
> - 3/5 use the top-to-bottom convention 👇 

I think that if you look at these examples the bottom-up and top-to-bottom conventions are used in different contexts:
- The 3 top-to-bottom examples are all examples of monthly values for a single calendar year. Arguably, it makes sense that January appears first when looking at a Jan-Dec table of "values". 
- On the other hand, when looking at arbitrary forecast horizons - as in the 2 bottom-up examples - I think that the bottom-up convention makes more sense since the lowest trace is the "closest" to the user. As the name indicated, a ridgeline plot can be interpreted by our eyes as a mountain ridgeline and in that sense it's more intuitive that the lowest trace is the closest and therefore the first.

Another argument for the bottom-up convention is that values should increase upwards on the y-axis.

## other comments

> _other option: left/right? might also be confusing, but time is typically left/right._

I saw some examples of this but they all looked very "busy". It's a personal preference but I did not like them. Violin plots could be an option but I think they would look even busier when looking at forecast horizons longer than a handful of units.

---

> _The orientation could be an option, with top-down default?_

Could be... but maybe setting a standard could also be a good idea here? If this becomes a more popular method of visualising probabilistic forecasts, we wouldn't won't to be in a situation where we always have to check the y-axis for directionality. It should become as natural as the left-to-right orientation of time-series plots.

---

> _we also should consider the case where the row dimension is not time, simply a sample index, like in sktime. E.g., a user might like to compare ridgeplots for group analysis or similar._

I need to see an example of this.

---

> _colours: Not sure whether I understand what's going on with the colour coding here. Using colours might help in general, can you explain?_

This is just the default behaviour from the ridgeplot library. The convention here could be anything we won't. It it's not currently supported by the ridgeplot library, I could try adding support. Currently, we accept [4 colormodes](https://ridgeplot.readthedocs.io/en/stable/api/public/ridgeplot.ridgeplot.html#ridgeplot.ridgeplot.params.colormode) for any arbitrary [colorscale](https://ridgeplot.readthedocs.io/en/stable/api/public/ridgeplot.ridgeplot.html#ridgeplot.ridgeplot.params.colorscale) (or you can use one of [Plotly's built-in named colorscales](https://plotly.com/python/builtin-colorscales/)).

---

> _I was wondering whether 5/25/50/75/95 quantiles could be connected by straights, whether that would make things look more "connected", or more confusing._

This could be done. I'll try working on an example. 🚧 

---

> _a substantially different plot - not ridgeplot - I was wondering about is left/right, plotting multiple quantiles, 5/25/50/75/95. That tends to look overcrowded though, and is not truly a distribution plot._

Something like this?

<img width="400" alt="image" src="https://github.com/tpvasconcelos/ridgeplot/assets/17701527/7897df83-08af-44d4-93fb-e118a713e269">

**Reply by fkiraly** [reply]
> It should become as natural as the left-to-right orientation of time-series plots.

Perhaps, I would put against that that I feel strongly about top-bottom in non-temporal context, as it is the "table" direction; and I would agree with bottom-up being better in a time series context.

The pain point this causes is that the table-like distributions in `skpro` are the same object used in both contexts, and the `sktime` `pd.DataFrame`-s progress from top to bottom in temporal order.

>> we also should consider the case where the row dimension is not time, simply a sample index, like in sktime. E.g., a user might like to compare ridgeplots for group analysis or similar.

> I need to see an example of this.

All examples in the `skpro` intro notebook are non-temporal, see there. No plots though, but that should suffice to get a picture, hopefully?

> This is just the default behaviour from the ridgeplot library. The convention here could be anything we won't. It it's not currently supported by the ridgeplot library, I could try adding support.

Would the colour shift be possible, from example number 5 below? That is, the colour at every point on the x-axis is the same above it.

**Reply by tpvasconcelos** [reply]
> Would the colour shift be possible, from example number 5 below? That is, the colour at every point on the x-axis is the same above it.

I would love to be able to do that as well but unfortunately Plotly does not have an option for filling curves with color gradients. See https://github.com/plotly/plotly.py/issues/3026 and https://github.com/plotly/plotly.js/issues/1918

**Reply by tpvasconcelos** [reply]
> All examples in the skpro intro notebook are non-temporal, see there. No plots though, but that should suffice to get a picture, hopefully?

Sounds good but maybe we can leave this as an exploration option for later on? Keeping my free time in mind, I think it would be best to start with a narrower scope.

I'll try refining my current implementation and example and:

- Add trace lines for a given list of quantiles (e.g. 5/25/50/75/95).
- Add support for panel data (more than one target for y)
- Showcase some examples for the different ways we could colour the different traces.

**Reply by fkiraly** [reply]
> Something like this?

yes, that looks brilliant!

**Reply by fkiraly** [reply]
>> All examples in the skpro intro notebook are non-temporal, see there. No plots though, but that should suffice to get a picture, hopefully?

> Sounds good but maybe we can leave this as an exploration option for later on? Keeping my free time in mind, I think it would be best to start with a narrower scope.

Sure - just noting, that the `skpro.distributions` module contains a carbon copy of the `sktime` one, it is a refactor target. So we need to be mindful that anything that works for temporally ordered collections for distributions also makes sense for unordered or grouped collections.

**Reply by tpvasconcelos** [reply]
> > Would the colour shift be possible, from example number 5 below? That is, the colour at every point on the x-axis is the same above it.
> 
> I would love to be able to do that as well but unfortunately Plotly does not have an option for filling curves with color gradients. See [plotly/plotly.py#3026](https://github.com/plotly/plotly.py/issues/3026) and [plotly/plotly.js#1918](https://github.com/plotly/plotly.js/issues/1918)

This is actually possible now thanks to the new `fillgradient` option for scatter plots.

I'm adding this feature in https://github.com/tpvasconcelos/ridgeplot/pull/244

**Comment by tpvasconcelos** [comment]
# A few examples and references that use this method to visualise probabilistic forecasts

This includes both univariable and panel series examples.

Granted that this is not a very significant sample but just for reference: 
- 2/5 use the bottom-up convention 👆 
- 3/5 use the top-to-bottom convention 👇 

---

<img width="410" alt="image" src="https://github.com/tpvasconcelos/ridgeplot/assets/17701527/ffb71735-ffee-43e2-a518-efe555013bad">

👉 https://seita.nl/2019/10/22/how-to-plot-probabilistic-forecasts/

---

<img width="600" alt="image" src="https://github.com/tpvasconcelos/ridgeplot/assets/17701527/2e504153-3f0d-49b0-bdd0-ef2032b33e51">

👉 https://link.springer.com/article/10.1007/s11222-023-10212-8

---

<img width="600" alt="image" src="https://github.com/tpvasconcelos/ridgeplot/assets/17701527/4bf7131a-2726-4182-bad3-f437a604e8e2">

👉 https://www.reddit.com/r/dataisbeautiful/comments/o765tz/oc_vaccination_rollout_by_age_belgium_ridgeline/

People seemed to have liked this way of visualising probabilistic forecasts. Here the non-OP top comment:

<img width="600" alt="image" src="https://github.com/tpvasconcelos/ridgeplot/assets/17701527/8af895c3-5d4d-445b-bce9-c6142ce2085f">

---

<img width="733" alt="image" src="https://github.com/tpvasconcelos/ridgeplot/assets/17701527/bc5a98ba-6c0e-43b3-a1f0-1b61061a3213">

---

<img width="460" alt="image" src="https://r-graph-gallery.com/img/graph/294-basic-ridgeline-plot3.png">

👉 https://wilkelab.org/ggridges/articles/introduction.html#varying-fill-colors-along-the-x-axis

---

**Reply by fkiraly** [reply]
Interesting, very interesting. My major takeaways would be:

* we are on the right track re plotting modality. People seem to find this way of plotting natural and informative.
* there are good reasons for both top/bottom and bottom-up. I think this should be an option.

Minor takeaways:

* the gradient in the last plot looks more natural than homogenous colour left/right.

**Reply by tpvasconcelos** [reply]
> the gradient in the last plot looks more natural than homogenous colour left/right.

I would love to be able to do that as well but unfortunately Plotly does not have an option for filling curves with color gradients. See https://github.com/plotly/plotly.py/issues/3026 and https://github.com/plotly/plotly.js/issues/1918

**Reply by tpvasconcelos** [reply]
> there are good reasons for both top/bottom and bottom-up. I think this should be an option.

Agreed!

**Reply by fkiraly** [reply]
> I would love to be able to do that as well but unfortunately Plotly does not have an option for filling curves with color gradients. 

I'm not that well-experienced in plotting, so I can't say what is easy or not. I agree with your general rationale - which I seem to be reading between your lines - that we should prioritize low-hanging fruit, of which there are plenty.

(example 5 is R, that's why they can do it?)

**Reply by fkiraly** [reply]
(random idea - not sure if it works in plotly: (a) filling with single colour or "transparent" placeholder; (b) layering with a large rectangle that has a gradient, where the colour in (a) acts as a "hole" in the plot)

**Reply by tpvasconcelos** [reply]
> (example 5 is R, that's why they can do it?)

Yes example 5 uses [ggridges](https://github.com/wilkelab/ggridges)

**Reply by tpvasconcelos** [reply]
This is now possible thanks to the new `fillgradient` option for scatter plots.

I'm adding this feature in https://github.com/tpvasconcelos/ridgeplot/pull/244

**Comment by fkiraly** [comment]
Some design questions:

* how should this be accessed? `BaseDistribution.plot`, with params?
* `ridgeplot` would have to become a managed soft dependency, e.g., in a plot softdep set
* the `statsmodels` dependency is a bit steep and not necessary for plots where you have access to `pdf`, `cdf`, etc. How do we manage that? Can we isolate it in `ridgeplot` proper? How about making the KDE backend a choice, and moving `statsmodels` to softdep in `ridgeplot` proper?

**Reply by fkiraly** [reply]
what's next, @tpvasconcelos?

**Comment by fkiraly** [comment]
@tpvasconcelos, here's a PR with a plotting interface, for probabiliy distributions in `skpro`.
https://github.com/sktime/skpro/pull/275

I was thinking about - and have designed it in that way - to allow different backends for plotting.

So, if used in a time series context, the distributions could automatically be switched to temporal plotting mode, why the panel mode makes more sense for tabular datasets where distributions are in no particular order.

**Comment by tpvasconcelos** [comment]
@fkiraly sorry for the radio silence. I lost the code I wrote to produce this initial draft

![image](https://github.com/user-attachments/assets/b988abf8-2c98-4a68-985e-d2ae88bff273)

which took me some time to get working. I will try to take a look at this again. I think that now that we have the gradient fill that you liked, the plots could look much nicer too :)

**Reply by fkiraly** [reply]
nice! Sorry, I stopped tracking the discussion due to the radio silence.

* do you have the code in a running version, or available as a prototype?
* would you like to present `ridgeplot` at one of the `sktime` meetups? We now record and put on youtube (if you like)

In case you lost the sktime discord link, here it is: https://discord.com/invite/54ACzaFsn7
Meet-ups are Fridays, 13 UTC.
